### PR TITLE
update RANSAC, ratio of inliers

### PR DESCRIPTION
**Pull request by Borda**
## Description
Adding option to use some init parameters for model fitting and RANDSAC initial inliers, eg, you have some a priory knowledge, you do not want to use some default parameters.

**NOTE:** this is copy of last version PR #2371 (because I had problems with rebase and made some mistakes)

## Checklist
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Unit tests

### CODE REVIEW

File: skimage/measure/fit.py [hunk 1]
```diff
@@ -341,13 +338,16 @@ class CircleModel(BaseModel):
...
+    def estimate(self, data, init_params=None):
         """Estimate circle model from data using total least squares.
 
         Parameters
         ----------
         data : (N, 2) array
             N points with ``(x, y)`` coordinates, respectively.
+        init_params : tuple of 3 values, optional
+            (x_center, y_center, r) is initial guess of parameters.
+            If they are not specified initial guess use a circle model.
```
Comment (sciunto) [inline skimage/measure/fit.py]: If None, use a circle with parameters based on mean values.

File: skimage/measure/fit.py [hunk 2]
```diff
@@ -479,13 +485,16 @@ class EllipseModel(BaseModel):
...
+    def estimate(self, data, init_params=None):
         """Estimate circle model from data using total least squares.
 
         Parameters
         ----------
         data : (N, 2) array
             N points with ``(x, y)`` coordinates, respectively.
+        init_params : tuple of 5 values, optional
+            (x_center, y_center, a, b, theta) is initial guess of parameters.
+            If they are not specified initial guess use a circle model.
```
Comment (sciunto) [inline skimage/measure/fit.py]: If None, use a circle with parameters based on mean values.

File: skimage/measure/fit.py [hunk 3]
```diff
@@ -378,14 +378,21 @@ def Dfun(params):
             A[0, :] = -(x - xc) / d
             A[1, :] = -(y - yc) / d
             # same for all iterations, so not changed in each iteration
-            #A[2, :] = -1
+            # A[2, :] = -1
```
Comment (sciunto) [inline skimage/measure/fit.py]: Why this commented line?

File: skimage/measure/fit.py [hunk 4]
```diff
@@ -378,14 +378,21 @@ def Dfun(params):
...
-        yc0 = y.mean()
-        r0 = dist(xc0, yc0).mean()
-        params0 = (xc0, yc0, r0)
-        params, _ = optimize.leastsq(fun, params0, Dfun=Dfun, col_deriv=True)
+        if init_params is None:
+            # initial guess of parameters using a circle model
+            xc0 = x.mean()
+            yc0 = y.mean()
+            r0 = dist(xc0, yc0).mean()
+            init_params = (xc0, yc0, r0)
```
Comment (sciunto) [inline skimage/measure/fit.py]: Concatenate these 4 lines into a single one.

File: skimage/measure/fit.py [hunk 5]
```diff
@@ -855,17 +872,21 @@ def ransac(data, model_class, min_samples, residual_threshold,
...
-    
+
     random_state = check_random_state(random_state)
 
-    if min_samples < 0:
+    if isinstance(min_samples, float):
+        if not (0 < min_samples <= 1):
+            raise ValueError("`min_samples` as ration must be in range (0, 1)")
+        min_samples = int(min_samples * len(data))
+    if min_samples <= 0:
```
Comment (sciunto) [inline skimage/measure/fit.py]: This is inconsistent with the docstring (zero included).

File: skimage/measure/fit.py [hunk 6]
```diff
@@ -770,7 +786,8 @@ def ransac(data, model_class, min_samples, residual_threshold,
         If RandomState instance, random_state is the random number generator;
         If None, the random number generator is the RandomState instance used
         by `np.random`.
-    
+    init_inliers : [list, tuple of] (N) array of bool or None, optional
+        Initial samples selection for model estimation
```
Comment (sciunto) [inline skimage/measure/fit.py]: estimation.

File: skimage/measure/fit.py (hunk 3 above)
Comment (Borda) [inline skimage/measure/fit.py]: no clue, it was there already before...

Comment (sciunto) [inline skimage/measure/fit.py]: Got it. Please rewrite the comment as 
```
# A[2, :] = -1 is the same for all iterations, not re-affected.
```

**Comment by sciunto** [comment]
Can you squash your commits and then edit the commit message to be explicit, ie `add init params for Ellipse & Circle and RANSAC`

**Comment by Borda** [comment]
squash commits is nice feature I have not used before... :)

File: skimage/measure/fit.py [hunk 7]
```diff
@@ -855,17 +868,21 @@ def ransac(data, model_class, min_samples, residual_threshold,
     best_inlier_num = 0
     best_inlier_residuals_sum = np.inf
     best_inliers = None
-    
+
     random_state = check_random_state(random_state)
 
+    if isinstance(min_samples, float):
+        if not (0 < min_samples <= 1):
```
Comment (sciunto) [inline skimage/measure/fit.py]: 0 <= min_samples <= 1

**Comment by sciunto** [comment]
looks correct to me, but I'm not using this feature. Anyone to review this PR?

**Comment by Borda** [comment]
Does anyone have an idea why it fails with "The build has been terminated" even on my personal branch [Borda / scikit-image](https://travis-ci.org/Borda/scikit-image/builds/186829586) it passes?

**Comment by soupault** [comment]
@Borda I guess it just got overwhelmed :).

**Comment by Borda** [comment]
@soupault @sciunto do you have anything else to this review?

**Comment by ahojnnes** [comment]
I have some concerns regarding this contribution:

1) It is not straightforward what to do when there are a priori inliers in RANSAC. The chosen strategy is certainly not "optimal" and could simply be replaced by 2 or 3 lines of code manually, if it is really needed. I am against adding this feature.
2) The documentation should motivate why `init_params` is useful (non-linear, iterative estimation). Currently, it is not clear what the purpose of the initialization is.

**Comment by Borda** [comment]
@ahojnnes regarding the 2) you are talking about description of `init_params` for RANSAC or also Circle? And do you mean adding just a short description or also an example?

**Comment by ahojnnes** [comment]
@Borda I mean to extend the doc string for `init_params` with a short one sentence description why a good set of starting parameters is useful (for convergence...).

**Comment by Borda** [comment]
it seems it crashs on building doc for `scikit-image/doc/source/api/skimage.filters.rank.rst`

**Comment by sciunto** [comment]
random failures due to travis. No worries.

**Comment by Borda** [comment]
@sciunto @ahojnnes any update? :)

File: skimage/measure/tests/test_fit.py [hunk 8]
```diff
@@ -1,4 +1,5 @@
 import numpy as np
+from scipy import spatial
```
Comment (stefanv) [inline skimage/measure/tests/test_fit.py]: Why is spatial added here?

**Comment by Borda** [comment]
it seems that there are some errors which came with merge upstream/master

* Cython.Compiler.Errors.CompileError: C:\projects\scikit-image\skimage\measure\_ccomp.pyx
* The executable /usr/bin/python3.6 (from --python=/usr/bin/python3.6) does not exist

**Comment by soupault** [comment]
I think it's time to let this PR go :smile: .
@Borda maybe try to select the related commits from the list in https://github.com/scikit-image/scikit-image/pull/2371/commits ? We could then make a rebase onto `master` and cherry pick them.

**Comment by Borda** [comment]
May I ask what happens with Travis CI, I am not able to build even the master...
https://travis-ci.org/Borda/scikit-image/builds/321960550

**Comment by soupault** [comment]
@Borda your `master` diverges from upstream one:
```
/home/travis/.travis/job_stages: line 57: tools/travis/before_install.sh: No such file or directory
```

**Comment by Borda** [comment]
ok, this seems to be solved...
but there appears an issue with space in doctest such as (this is from travis)
```
381     >>> np.round(ellipse.params, 2)
Expected:
    array([ 10.  ,  15.  ,   4.  ,   8.  ,   0.52])
Got:
    array([10.  , 15.  ,  4.  ,  8.  ,  0.52])
```

**Comment by soupault** [comment]
@Borda yet, that is known. See https://github.com/scikit-image/scikit-image/pull/2935

**Comment by Borda** [comment]
I see I was thinking about correcting these test with correcting the output and set `# doctest: +NORMALIZE_WHITESPACE` and do it as new PR from master... what do you think?

**Comment by soupault** [comment]
From what I understood, it is not just a matter of whitespaces - see https://github.com/numpy/numpy/blob/master/doc/release/1.14.0-notes.rst#many-changes-to-array-printing-disableable-with-the-new-legacy-printing-mode . Floating point precision could be different, `bool` ndarrays dtype is not printed anymore, etc.

**Comment by Borda** [comment]
yes, I noticed it also while I tried to solve it in #2938, but the difficulty comes with the configurations because in some configurations seems to be used the old formatting...

**Comment by Borda** [comment]
It seems that I lost the commits... :-/

**Pull request closed without merging**